packages/keystone/plugins/softDeleted.js

```diff
@@ -5,6 +5,45 @@ const { composeResolveInputHook, evaluateKeystoneAccessResult } = require('./uti
 const { plugin } = require('./utils/typing')
 
 const { queryHasField } = require('../queryHasField')
+const { find, getListDependentRelations } = require('../schema')
+
+const PROTECT = 'models.PROTECT'
+const SET_NULL = 'models.SET_NULL'
+const CASCADE = 'models.CASCADE'
+
+const hasObjs = async (schemaName, path, objId) => {
+    // BillingReceipt { path: objId }
+    const where = { [path]: { id: objId }, deletedAt: null }
+
+    // If there are any objects that have this ID
+    return await find(schemaName, where)
+}
+
+const canDelete = async (listName, obj) => {
+    const relations = getListDependentRelations(listName)
+
+    const objId = typeof obj === 'string' ? obj : obj.id
+
+    for (const rel of relations) {
+        const onDelete = get(rel, ['config', 'kmigratorOptions', 'on_delete'])
+
+        if (onDelete === PROTECT) {
+            const existingDependants = await hasObjs(rel.from, rel.path, objId)
+            if (existingDependants.length > 0) {
+                throw new Error(`You can not delete ${rel.to}:${objId}, dependant: ${rel.from} exists: ${existingDependants.map(x => x.id).join(',')}, and on_delete rule on ${rel.from} set to ${rel.onDelete}`)
```

Comment:
Why not GQLError?

Reply:
It's a bit out of infra team style, but any tests?)

Reply:
At first we need to check out all other models and fix PROTECT relation rules inside them. I think this is going to be done in this PR :-)